### ISSUE TITLE
spinful_fermion_basis_1d does not work for L>32

### ISSUE THREAD

**Issue by anton-buyskikh** [weinbe58/QuSpin#242]
I'm using
```bash
quspin 0.3.2
numpy 1.16.4
```

This code
```python
from quspin.basis import spinful_fermion_basis_1d
basis = spinful_fermion_basis_1d(L, Nf=(N_up, N_dn))
```
produces the following error
```bash
Traceback (most recent call last):
  File "quspin/basis/basis_1d/_basis_1d_core/sources/spf_next_state.pyx", line 19, in quspin.basis.basis_1d._basis_1d_core.spf_basis.next_state_pcon_spf
  File "quspin/basis/basis_1d/_basis_1d_core/sources/spf_next_state.pyx", line 8, in quspin.basis.basis_1d._basis_1d_core.spf_basis.next_state_pcon
TypeError: unsupported operand type(s) for >>: 'float' and 'int'
Exception ignored in: 'quspin.basis.basis_1d._basis_1d_core.spf_basis.make_n_basis_template'
Traceback (most recent call last):
  File "quspin/basis/basis_1d/_basis_1d_core/sources/spf_next_state.pyx", line 19, in quspin.basis.basis_1d._basis_1d_core.spf_basis.next_state_pcon_spf
  File "quspin/basis/basis_1d/_basis_1d_core/sources/spf_next_state.pyx", line 8, in quspin.basis.basis_1d._basis_1d_core.spf_basis.next_state_pcon
TypeError: unsupported operand type(s) for >>: 'float' and 'int'
```
for `L > 32`, `N_up > 0`, and any `N_dn`.

On the other hand `spinful_fermion_basis_general` works fine.

**Comment by mgbukov** [comment]
I was able to reproduce the bug. It arises in the expression `(t & -t) / (v & -v)` which for L>32 results in a float even though both `t` and `v` are `basis_type`. 

@weinbe58 any idea why cython behaves this way? I thought it's strongly typed. Btw, you must've encountered this bug before because the corresponding expression for the `hpc` basis contains the integer division `(t & -t) // (v & -v)` which removes the bug. If all is fine, go ahead and merge the pull request I just created. 

This is now fixed and a pull request is created into `user_basis`. We'll close the issue with the new release.
